RNFtools
========

.. image:: https://travis-ci.org/karel-brinda/rnftools.svg?branch=master
	:target: https://travis-ci.org/karel-brinda/rnftools

.. image:: https://readthedocs.org/projects/rnftools/badge/?version=latest
	:target: http://rnftools.rtfd.org

.. image:: https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat-square
	:target: https://anaconda.org/bioconda/rnftools

.. image:: https://badge.fury.io/py/RNFtools.svg
	:target: https://badge.fury.io/py/RNFtools

.. image:: https://img.shields.io/badge/doi-10.1093%2Fbioinformatics%2Fbtv524-brightgreen.svg
	:target: http://dx.doi.org/10.1093/bioinformatics/btv524

.. image:: https://zenodo.org/badge/DOI/10.5281/zenodo.1066626.svg
	:target: https://doi.org/10.5281/zenodo.1066626

**Read Naming Format** is a generic format for assigning
read names with encoded information about original positions.
**RNFtools** is an associated
software package which can:

* simulate RNF-compliant reads using a wide class of integrated read simulators (Art, DwgSim, Mason, WgSim, etc.), either from a single genome (i.e., a whole genome sequencing) or multiple genomes (i.e., a metagenomic simulation);
* evaluate mappers using RNF reads;
* convert non-RNF simulated reads to RNF (e.g., from SAM format);
* transform genomic coordinates of RNF reads between different coordinate systems (using chain LiftOver format).

Links
-----

**Web of the project:** http://karel-brinda.github.io/rnftools/

**RNF specification:** http://karel-brinda.github.io/rnf-spec/

**Documentation:** http://rnftools.rtfd.org

**Installation:** http://rnftools.readthedocs.org/en/latest/tutorial/00_installation.html

**Examples of usage:** http://github.com/karel-brinda/rnftools/tree/master/examples/01_tutorial

**Publication:** http://dx.doi.org/10.1093/bioinformatics/btv524
